### PR TITLE
[asset-hubs/bridge-hubs] Bridging Polkadot Asset Hub <-> Kusama Asset Hub to enable asset transfer of KSMs/DOTs

### PR DESCRIPTION
**Pull request by bkontur**
## Summary

Adding bridging capabilities to the Polkadot and Kusama Bridge Hub runtimes and XCM routing configuration for bridging between Polkadot Asset Hub <-> Kusama Asset Hub in order to allow asset transfer between Asset Hubs.

## Motivation

Main goal of these changes is to setup bridge between system parachains Kusama Asset Hub vs Polkadot Asset Hub. We explicitly allow only reserve based transfer for KSM/DOT to cross-consensus AssetHubs and nothing else at first (start defensively). We can later allow whatever (ETH, TrustBackedAssets, ForeignAssets, PoolAssets ...).

## Explanation

Please, check first four commits with detailed description

## Testing

Adds tests for simulation:
- `limited_reserve_transfer_assets` with `pallet_xcm` both ways
- handling `ReserveAssetDeposisted` on both sides
- handling `ExportMessage` on both sides
- handling message dispatch and routing to target on both side
- relayer can submit proofs with `submit_finality_proof` / `submit_parachain_heads` and relay messages with `receive_messages_proof`

Adds local zombienet run for:
- reserve based transfer of KSMs from AHK to AHP
- reserve based transfer of DOTs from AHP to AHK

*(check parachains/runtimes/bridge-hubs/README.md in this PR)*

## Future Directions and Related Material

Next stage will be focused on adding support to allow bridging between arbitrary parachains More about [Bridges V2](https://github.com/paritytech/parity-bridges-common/milestone/17).

More about dynamic fees and back-preasure for v1 can be found [here](https://github.com/paritytech/parity-bridges-common/pull/2294).

This is a migration (without commit history) of original PRs in cumulus:
- [x] https://github.com/paritytech/cumulus/pull/2762 (cherry-picked from @acatangiu's https://github.com/paritytech/polkadot-sdk/pull/1224)
- [x] https://github.com/paritytech/cumulus/pull/2997
- [x] https://github.com/paritytech/cumulus/pull/2528
- [x] https://github.com/paritytech/cumulus/pull/3001

## TODO
- [x] finish migration
- [x] fix `cumulus/parachains/runtimes/bridge-hubs/README.md` for local run
- [x] update description (collect all relevant information from migrated PRs)
- [ ] add integration tests with xcm-emulator
- [ ] migrate/open PR with runtime modifications to `polkadot-fellows` https://github.com/paritytech/parity-bridges-common/issues/2541

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3543585

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3543742

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3549836

**Review by svyatonik — APPROVED** [review]

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-polkadot/src/bridge_hub_config.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Bridge definitions.
```
Comment (acatangiu) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-polkadot/src/bridge_hub_config.rs:17]:
```suggestion
//! Polkadot BridgeHub definitions.
```

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-polkadot/src/bridge_hub_config.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	RefundableMessagesLane<WithBridgeHubKusamaMessagesInstance, StatemintToStatemineMessageLane>,
+	ActualFeeRefund<Runtime>,
+	PriorityBoostPerMessage,
+	StrBridgeRefundBridgeHubKusamaMessages,
+>;
+bp_runtime::generate_static_str_provider!(BridgeRefundBridgeHubKusamaMessages);
+
+// TODO: rework once dynamic lanes are supported (https://github.com/paritytech/parity-bridges-common/issues/1760)
+//       now we support only StatemineToStatemint
+/// Lanes setup
```
Comment (acatangiu) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-polkadot/src/bridge_hub_config.rs:172]:
```suggestion
//       now we support only AHP<>AHK Lanes setup
```

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-polkadot/src/xcm_config.rs [hunk 3]
```diff
@@ -129,6 +134,17 @@ impl Contains<RuntimeCall> for SafeCallFilter {
 			}
 		}
 
+		// Allow to change dedicated storage items (called by governance-like)
+		match call {
+			RuntimeCall::System(frame_system::Call::set_storage { items })
+				if items.iter().any(|(k, _)| {
```
Comment (acatangiu) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-polkadot/src/xcm_config.rs:140]:
```suggestion
				if items.iter().all(|(k, _)| {
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3567053

File: cumulus/parachains/common/src/xcm_config.rs [hunk 4]
```diff
@@ -78,3 +86,194 @@ impl<Location: Get<MultiLocation>> ContainsPair<MultiAsset, MultiLocation>
...
+						.add_equals(X1(Parachain(2000)))
+						.add_equals(X1(Parachain(3000)))
+						.add_equals(X1(Parachain(4000))),
+					BridgeLocation::get(),
+					None
+				)
+			];
+		}
+
+		let test_data = vec![
```
[the remark this answers is not in the document]
Comment (bkontur) [reply cumulus/parachains/common/src/xcm_config.rs:255]: there are three `Some(_)` expected results - lines 262-264:
```
			(Kusama, X1(Parachain(2000)), Some((BridgeLocation::get(), None))),
			(Kusama, X1(Parachain(3000)), Some((BridgeLocation::get(), None))),
			(Kusama, X1(Parachain(4000)), Some((BridgeLocation::get(), None))),
```
or what do you exactly mean?

[the remark this answers is not in the document]
Comment (acatangiu) [reply cumulus/parachains/common/src/xcm_config.rs:255]: oh, right, my eyes deceived me, it's good! 👍

**Review by acatangiu — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3570651

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3571145

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-kusama-bridge/2971/5

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3739791

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3760830

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3761486

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3761745

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3770838

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3770968

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3770998

**Comment by acatangiu** [comment]
closing in favor of https://github.com/polkadot-fellows/runtimes/pull/108

**Pull request closed without merging**